### ISSUE TITLE
Cannot use with webpack

### ISSUE THREAD

**Issue by krojew** [melonjs/melonJS#832]
I tried packaging my app with melon bundled in using Webpack and ultimately failed:

WARNING in ./XXX/melonJS.js
Module not found: Error: Cannot resolve module 'vertx' in /XXX

So I started to investigate and it turns out melon packs some polyfills, especially es6-promise. That one tries different methods to implement promises and one of them tries to require vertx, which might not be present. Webpack sees this 'require' call and tries to resolve the module and we have an error. The solution might be using something more modern, like core-js.

**Comment by agmcleod** [comment]
@krojew What loaders are you using with webpack? My current project uses with babel, but i avoid using the polyfill. To save weight, and I know because we polyfill a couple things already. Though searching on the use of promises, I found none. @obiot do you know why it was added?

Here's my webpack config in case you'd like to see it for reference:

```
var path = require('path');
var webpack = require('webpack');

module.exports = {
  entry: [
    './lib/melonJS.js',
    './lib/plugins/debugPanel.js',
    './lib/spine.js',
    './lib/melonjs-spine.js',
    './lib/plugins/debugPanel.js',
    './js/game.js'
  ],
  output: {
    path: __dirname,
    filename: 'bundle.js'
  },
  watch: true,
  devtool: 'source-map',
  module: {
    loaders: [{
      test: /\.html/, loader: 'file-loader'
    }, {
      test: /\.js$/,
      include: path.join(__dirname, 'js'),
      loader: 'babel-loader',
      exclude: /lib/
    }, {
      test: /\.js$/,
      loader: 'eslint-loader',
      exclude: /node_modules|lib/
    }]
  }
};
```

As well as my package.json:

```{
"name": "boilerplate",
"version": "0.1.0",
"melon_version": "2.1.3",
"devDependencies": {
"babel-core": "^6.1.21",
"babel-eslint": "^4.1.5",
"babel-loader": "^6.2.0",
"babel-plugin-transform-runtime": "^6.1.18",
"babel-preset-es2015": "^6.1.18",
"copy-webpack-plugin": "^0.3.3",
"csv-parse": "^1.1.7",
"eslint": "^1.9.0",
"eslint-loader": "^1.1.1",
"file-loader": "^0.8.5",
"webpack": "^1.12.6",
"webpack-dev-server": "^1.12.1"
},
"dependencies": {
"babel-runtime": "^6.3.13"
},
"scripts": {
"parse_sections": "node ./bin/build_sections.js",
"start": "node node_modules/webpack-dev-server/bin/webpack-dev-server.js --port 8100",
"build": "node node_modules/webpack/bin/webpack.js --config webpack.production.config.js"
}
}```

**Comment by krojew** [comment]
For js I use babel-loader too. Have you tried webpack.optimize.CommonsChunkPlugin?

**Comment by agmcleod** [comment]
I havent, and i probably should. My build size is still pretty small, and im serving it through cordova more so than the web.

**Comment by obiot** [comment]
@agmcleod for the promise polyfill it's used by Howler, and then would be required for ticket #678

**Comment by obiot** [comment]
and guess what, IE11 does not support promises .... :P
http://caniuse.com/#feat=promises

**Comment by krojew** [comment]
Maybe consider switching to core-js?   This and switching to npm for dependency management would make melon highly usable with complex projects.

**Comment by obiot** [comment]
so i;m confused, is there anything need doing, or can this ticket be closed ?

**Comment by krojew** [comment]
The problem is still there. To solve it you need to do any of these things (preferably all of them):

- make melon possible to install via npm, which implies no baked-in dependencies in the package
- make a standalone package without dependencies baked-in, so people can fill them in using their toolchain of choice
- switch built-in dependencies to something more modern for the "full" package

**Comment by agmcleod** [comment]
I just want to be clear though, it's a loader that has the dependency resolution problem, not webpack. Webpack on its own does not check for any specific libraries or polyfills unless you tell it to with a loader.

But you're right, the solution here is to move it to a more npm based dependency structure.

**Comment by obiot** [comment]
so this is then redundant with #772.

**Comment by krojew** [comment]
@obiot yes and no. #772 is one of the possible fixes, the best one in my opinion. But there also are other alternatives. Yet, if you wish to fix it by #772 that's a good choice, since npm is now de facto a standard for JS libraries.

**Comment by parasyte** [comment]
IMHO, we can drop almost all of these polyfills. And that work has been started by #829 Given that and #772, I suppose this ticket is redundant.
